src/Sarif.Multitool/Rules/SARIF1021.MissingToolVersion.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class MissingToolVersion : SarifValidationSkimmerBase
+    {
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1021_MissingToolVersion
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        /// <summary>
+        /// SARIF1021
+        /// </summary>
+        public override string Id => RuleId.MissingToolVersion;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1021_Default)
+        };
+
+        protected override void Analyze(Tool tool, string pointer)
+        {
+            // 'Driver' is a required property, hence we do not need a null check for it. 
+            if (tool.Driver.Version == null)
```

Comment:
Here, you are just checking for a null value. The other cases that I can see:
- version is white space or empty: is that an error as well?
- version is not a version. For example, `tool.Driver.Version = "version"`: is that an error as well?

Reply:
great idea, i should add a check for whitespaces. reg the version scheme, yes that could be a potential test but i am worried if it will be too constraining. each tool is free to choose how they version themselves. 

---
In reply to: [434486864](https://github.com/microsoft/sarif-sdk/pull/1899#discussion_r434486864) [](ancestors = 434486864)

Reply:
Yes, each tool can choose its versioning scheme. It is only `semanticVersion` that is constrained.

I see you expanded the check to include empty or white space.

---
In reply to: [434737873](https://github.com/microsoft/sarif-sdk/pull/1899#discussion_r434737873) [](ancestors = 434737873,434486864)

Reply:
Reactivating. Eddy has a good idea for expanding this rule, which would be both to require the presence of a version and that it seems some sort of standard in terms of its contents. Note that this may require the rule to return multiple severities. It also raises the question of whether sub-analyses are covered by subrules. So probably a bit more discussion to do.

---
In reply to: [434753356](https://github.com/microsoft/sarif-sdk/pull/1899#discussion_r434753356) [](ancestors = 434753356,434737873,434486864)